### PR TITLE
AL/MIN-167. Liquidate loans fix.

### PR DESCRIPTION
**Pull request by lut1k**
**Description:**

Fix the logic in function liquidate_calculate_seize_and_repay().
If the liquidation pool does not have enough funds to pay off the debt, then the liquidation pool repays the amount of assets available to it. The number of liquidation attempts does not change.

**Changes in storage or API:**

--

Describe storage structure changes or API changes. Make sure to update the [API docs](https://minterestfinance.atlassian.net/wiki/spaces/MINTEREST/pages/134807557/Minterest+API).

**Checklist**

- [ ] Code is covered with tests
- [ ] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

File: runtime/src/tests/liquidation_tests.rs [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
+
+Collateral factor = 90% for all pools.
+Alice - depositer, Bob - borrower.
+1. Bob made a deposit into the system and set as collateral DOT and BTC.
+2. Bob borrowed ETH.
+3. Ethereum price decreased.
+4. The first partial liquidation.
+5. Bob withdrew all the collateral DOT.
+6. Bob withdrew BTC and left only 1 bitcoin in the protocol.
+7. Bitcoin price has increased.
```
Comment (DenisRomanovsky) [inline runtime/src/tests/liquidation_tests.rs:14]: Why increase of collateral value triggers a liquidation? May be a typo and it should be a decrease?

Comment (Apollo50) [reply runtime/src/tests/liquidation_tests.rs:14]: BTC isn't a Bob's collateral. In this scope Alice deposited BTC to add liquidity to BTC pool.

Comment (lut1k) [reply runtime/src/tests/liquidation_tests.rs:14]: Bob borrowed bitcoin. Bitcoin price has increased

File: runtime/src/tests/liquidation_tests.rs [hunk 2]
```diff
@@ -427,7 +432,163 @@ fn partial_liquidation_should_not_work() {
...
+			);
+
+			assert_eq!(
+				LiquidityPools::pool_user_data(DOT, ALICE::get()).liquidation_attempts,
+				3
+			);
+		})
+}
+
+// If the liquidation pool does not have enough funds to pay off the debt, then the
```
Comment (DenisRomanovsky) [inline runtime/src/tests/liquidation_tests.rs:486]: The three comments are repeated. Guess we can use one of them or modify for each case.

File: pallets/risk-manager/src/lib.rs [hunk 3]
```diff
@@ -613,24 +614,38 @@ impl<T: Config> Pallet<T> {
...
-			.checked_mul(&temporary_factor)
+		let liquidation_pool_balance = T::LiquidationPoolsManager::get_pool_available_liquidity(liquidated_pool_id);
+		let liquidation_pool_balance_usd = Rate::from_inner(liquidation_pool_balance)
+			.checked_mul(&price_borrowed)
+			.map(|x| x.into_inner())
+			.ok_or(Error::<T>::NumOverflow)?;
+
+		// If there is not enough liquidity in the liquidation pool, then we do not change
+		// the user's liquidation attempts counter.
+		let is_need_mutate_attempts = liquidation_pool_balance_usd >= repay_amount;
```
Comment (DenisRomanovsky) [inline pallets/risk-manager/src/lib.rs:641]:
```suggestion
		let increment_attempt_counter = liquidation_pool_balance_usd >= repay_amount;
```
Guess is_ prefix is not always suitable. is_empty is ok, is_increment_attempt_counter_required is good, but long.

File: pallets/risk-manager/src/lib.rs [hunk 4]
```diff
@@ -613,24 +614,38 @@ impl<T: Config> Pallet<T> {
...
 
 		let temporary_factor = match is_partial_liquidation {
 			true => Rate::saturating_from_rational(30, 100),
 			false => Rate::one(),
 		};
 
-		// seize_amount = liquidation_fee * temporary_factor * total_repay_amount
-		let seize_amount = Rate::from_inner(total_repay_amount)
+		// repay_amount = temporary_factor * total_repay_amount
+		let mut repay_amount = Rate::from_inner(total_repay_amount)
```
Comment (DenisRomanovsky) [inline pallets/risk-manager/src/lib.rs:628]: The annotation says this is the borrowed amount in usd, not the amount to be repayed. Either correct the annotation or the code.

File: runtime/src/tests/liquidation_tests.rs (hunk 1 above)
Comment (DenisRomanovsky) [reply runtime/src/tests/liquidation_tests.rs:14]: Line 2 of the scenario says Bob borrowed ETH and deposited BTC.

File: pallets/risk-manager/src/lib.rs [hunk 5]
```diff
@@ -474,12 +474,13 @@ impl<T: Config> Pallet<T> {
 			&& liquidation_attempts < RiskManagerParams::<T>::get(liquidated_pool_id).max_attempts;
 
 		// Calculate sum required to liquidate.
-		let (seize_amount, repay_amount, repay_assets) =
+		let (seize_amount, repay_amount, repay_assets, is_need_mutate_attempts) =
```
Comment (dev-vadim) [inline pallets/risk-manager/src/lib.rs:477]: It`s better to group these values in some struct. And maybe also rename liquidate_calculate_seize_and_repay

File: runtime/src/tests/liquidation_tests.rs (hunk 1 above)
Comment (lut1k) [reply runtime/src/tests/liquidation_tests.rs:14]: Fixed ><

File: runtime/src/tests/liquidation_tests.rs (hunk 2 above)
Comment (lut1k) [reply runtime/src/tests/liquidation_tests.rs:486]: Done

File: pallets/risk-manager/src/lib.rs (hunk 3 above)
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:641]: Fixed

File: pallets/risk-manager/src/lib.rs (hunk 4 above)
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:628]: Fixed annotation

File: pallets/risk-manager/src/lib.rs (hunk 5 above)
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs]: Check, please

File: pallets/risk-manager/src/lib.rs [hunk 6]
```diff
@@ -604,33 +620,43 @@ impl<T: Config> Pallet<T> {
...
-	pub fn liquidate_calculate_seize_and_repay(
+	/// Returns LiquidationInfo.
+	pub fn calculate_liquidation_info(
 		liquidated_pool_id: CurrencyId,
 		total_repay_amount: Balance,
 		is_partial_liquidation: bool,
-	) -> result::Result<(Balance, Balance, Balance), DispatchError> {
+	) -> result::Result<LiquidationInfo, DispatchError> {
 		let liquidation_fee = Self::risk_manager_dates(liquidated_pool_id).liquidation_fee;
+		let price_borrowed =
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:630]: Fixed

File: pallets/risk-manager/src/lib.rs [hunk 7]
```diff
@@ -604,33 +620,43 @@ impl<T: Config> Pallet<T> {
...
+		let mut repay_amount = Rate::from_inner(total_repay_amount)
 			.checked_mul(&temporary_factor)
-			.and_then(|v| v.checked_mul(&liquidation_fee))
 			.map(|x| x.into_inner())
 			.ok_or(Error::<T>::NumOverflow)?;
 
-		// repay_amount = temporary_factor * total_repay_amount
-		let repay_amount = Rate::from_inner(total_repay_amount)
-			.checked_mul(&temporary_factor)
+		let liquidation_pool_balance = T::LiquidationPoolsManager::get_pool_available_liquidity(liquidated_pool_id);
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:644]: Regenerate

**Review by dev-vadim — APPROVED** [review]

**Review by DenisRomanovsky — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**